src/rpc/tensorpipe/tp_communicator.h

```diff
@@ -0,0 +1,186 @@
+   * \brief Finalize TPSender
+   */
+  void Finalize();
+
+  /*!
+   * \brief Communicator type: 'tp'
+   */
+  inline std::string Type() const { return std::string("tp"); }
+
+ private:
+  /*!
+   * \brief global context of tensorpipe
+   */
+  std::shared_ptr<tensorpipe::Context> context;
+
+  /*!
+   * \brief pipe for each connection of receiver
+   */
+  std::unordered_map<int /* receiver ID */, std::shared_ptr<tensorpipe::Pipe>>
+    pipes_;
+
+  /*!
+   * \brief receivers' listening address
+   */
+  std::unordered_map<int /* receiver ID */, std::string> receiver_addrs_;
+};
+
+/*!
+ * \brief TPReceiver for DGL distributed training.
+ *
+ * Tensorpipe Receiver is the communicator implemented by tcp socket.
+ */
+class TPReceiver {
+ public:
+  /*!
+   * \brief Receiver constructor
+   * \param queue_size size of message queue.
+   */
+  explicit TPReceiver(std::shared_ptr<tensorpipe::Context> ctx) {
+    CHECK(ctx) << "Context is not initialized";
+    this->context = ctx;
+    queue_ = std::make_shared<RPCMessageQueue>();
+  }
+
+  /*!
+   * \brief Wait for all the Senders to connect
+   * \param addr Networking address, e.g., 'tcp://127.0.0.1:50051'
+   * \param num_sender total number of Senders
+   * \return True for success and False for fail
+   *
+   * Wait() is not thread-safe and only one thread can invoke this API.
+   */
+  bool Wait(const std::string& addr, int num_sender);
+
+  /*!
+   * \brief Recv RPCMessage from Sender. Actually removing data from queue.
+   * \param msg pointer of RPCmessage
+   * \param send_id which sender current msg comes from
+   * \return Status code
+   *
+   * (1) The Recv() API is blocking, which will not
+   *     return until getting data from message queue.
+   * (2) The Recv() API is thread-safe.
+   * (3) Memory allocated by communicator but will not own it after the function
+   * returns.
+   */
+  void Recv(RPCMessage* msg);
+
+  /*!
+   * \brief Finalize SocketReceiver
+   *
+   * Finalize() is not thread-safe and only one thread can invoke this API.
+   */
+  void Finalize();
+
+  /*!
+   * \brief Communicator type: 'tp' (tensorpipe)
+   */
+  inline std::string Type() const { return std::string("tp"); }
+
+  /*!
+   * \brief Issue a receive request on pipe, and push the result into queue
+   */
+  static void ReceiveFromPipe(std::shared_ptr<tensorpipe::Pipe> pipe,
+                              std::shared_ptr<RPCMessageQueue> queue);
+
+ private:
+  /*!
+   * \brief number of sender
+   */
+  int num_sender_;
+
+  /*!
+   * \brief listener to build pipe
+   */
+  std::shared_ptr<tensorpipe::Listener> listener;
+
+  /*!
+   * \brief global context of tensorpipe
+   */
+  std::shared_ptr<tensorpipe::Context> context;
+
+  /*!
+   * \brief pipe for each client connections
+   */
+  std::unordered_map<int /* Sender (virutal) ID */,
+                     std::shared_ptr<tensorpipe::Pipe>>
+    pipes_;
+
+  /*!
+   * \brief RPCMessage queue
+   */
+  std::shared_ptr<RPCMessageQueue> queue_;
```

Comment:
i try to understand how send and receive work after using TensorPipe. Do we still need to have a thread dedicated to waiting for messages? I see receiver has a method called Wait.

Reply:
Yes. Wait is at the connection stage, which is corresponding to sender's `Connect` function. It's not for waiting a RPC Message